### ISSUE TITLE
Amazon Lambda SAM CLI integration

### ISSUE THREAD

**Issue by oztimpower** [quarkusio/quarkus#5037]
**Description**
I've tested the SAM integration that is on the Amazon Lambda HTTP docs, and used a similar sam.jvm.yaml (without the dynamo db YAML), and it works fine.

Any chance the archetype and documentation could clone this functionality from the Amazon Lambda HTTP implementation (incl. uber jar) -- particularly whilst it's fresh in the minds of the developers.

```
./extensions/amazon-lambda-http/maven-archetype/target/classes/archetype-resources/sam.jvm.yaml
./extensions/amazon-lambda-http/maven-archetype/target/classes/archetype-resources/sam.native.yaml
./extensions/amazon-lambda-http/maven-archetype/src/main/resources/archetype-resources/sam.jvm.yaml
./extensions/amazon-lambda-http/maven-archetype/src/main/resources/archetype-resources/sam.native.yaml

```

```
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Globals:
  Function:
    Timeout: 5
Resources:
  QuarkusSam:
    Type: AWS::Serverless::Function
    Properties:
      Handler: io.quarkus.amazon.lambda.runtime.QuarkusStreamHandler::handleRequest
      Runtime: java8
      CodeUri: target/lambda-kotlin-1.0-SNAPSHOT-runner.jar
      MemorySize: 128
      Policies: AWSLambdaBasicExecutionRole

```

It would be super cool also to have this in Gradle which is my preferred build tool.

**Comment by geoand** [comment]
cc @evanchooly @patriot1burke

**Comment by evanchooly** [comment]
I'm not quite sure I understand what you're asking here.  Apart from the gradle request.  I can talk to the guy doing the gradle integration for the most appropriate route there.

**Comment by oztimpower** [comment]
Primary request is to auto-template the sam.jvm.yaml and sam.native.yaml similar to the amazon-lambda-http extension, which is what I'm describing in the amazon-lambda guide (as per your feedback).

Gradle request was secondary.

**Comment by evanchooly** [comment]
oh, right, because `amazon-lambda` isn't currently set up for those templates where `amazon-lambda-http` is.  Got it.

**Comment by evanchooly** [comment]
So i've updated the `amazon-lambda` to provide these templates and improved the scripts generated to be in sync with the template files.  I've pushed to my fork at https://github.com/evanchooly/quarkus/tree/lambda-sam.  If you want to look it over and see if that's what you had in mind, I can work on a PR once you're done.  The templates should be all but identical to the `amazon-lambda-http` templates but I did do some clean up on the shell scripts.

**Comment by oztimpower** [comment]
The simplification via manage.sh and the lambda role env. variable is great to see.

I quite liked the detail and script content on the former documentation, it is a shame to remove all of that; it allowed me to construct Gradle equivalents for example, without having to generate the maven archetype. 

The sam.jvm.yaml and sam.native.yaml template still have the API Gateway detail in the template, which needs to be stripped for a standard Amazon Lambda function. In the documentation I had both the jvm and native examples of the reduced template, that removed the API Gateway related configuration.

**Comment by oztimpower** [comment]
Tested it, works fine.  Nice one!

Is there a way to make manage.sh executable via the archetype rather than having to use sh to invoke it? i.e., ./manage.sh

Since there is also a sample payload.json, the sam docs could be updated to refer to it?
`sam local invoke --template sam.native.yaml --event payload.json `
  
ProcessingService.process has a typo (it's currently proces).

Docs look great simplified, ignore my comment earlier, it's easy to use the maven archetype, even if you have a Gradle bias :)

**Comment by evanchooly** [comment]
I just put up a draft PR if you want to take one last look before I submit it.

**Comment by oztimpower** [comment]
Templates look much simpler. There is still some API Gateway config in it, which can be removed:

  Globals:
     Api:
       EndpointConfiguration: REGIONAL
       BinaryMediaTypes:
         - "*/*"

**Comment by oztimpower** [comment]
In the docs, would it be better to change that paragraph to always add the -D param, since that's consistent with the other docs on building for a container.  It would actually be nice to have this written as code block like:
`mvn -Pnative -Dnative-image.docker-build=true`

===>
If you want a lower memory footprint and faster initialization times for your lambda, you can compile your Java  code to a native executable.  Just make sure to rebuild your project with the `-Pnative` switch.  If you are building on a non-linux system, you will need to also pass in a property instructing quarkus to use a docker build as Amazon
 Lambda requires linux binaries.  You can do this by passing this property to your maven build:
 `-Dnative-image.docker-build=true`.

**Comment by oztimpower** [comment]
Majority of the macro code in the template can go?  Certainly all the Region related stuff can.

#set($resourceName = $artifactId)
 #macro(replaceChar $originalName, $char)
 #if($originalName.contains($char))
 #set($tokens = $originalName.split($char))
 #set($newResourceName = "")
 #foreach($token in $tokens)
 #set($newResourceName = $newResourceName + $token.substring(0,1).toUpperCase() + $token.substring(1).toLowerCase())
 #end
   ${newResourceName}
   #else
   #set($newResourceName = $originalName.substring(0,1).toUpperCase() + $originalName.substring(1))
   ${newResourceName}
   #end
   #end
   #set($resourceName = "#replaceChar($resourceName, '-')")
   #set($resourceName = "#replaceChar($resourceName, '.')")
   #set($resourceName = $resourceName.replaceAll("\n", "").trim())
   #macro(regionVar)
   AWS::Region
 #end
 #set($awsRegion = "#regionVar()")
 #set($awsRegion = $awsRegion.replaceAll("\n", "").trim())

**Comment by oztimpower** [comment]
Congrats, looks super simple and easy to follow.  It's a big improvement, allowing people to test out of the box locally.  No other feedback!

**Comment by patriot1burke** [comment]
So, I want the docs to describe the AWS CLI commands that are supposed to run.   Not sure why you wanted to take that out.

**Comment by patriot1burke** [comment]
and why do you want to strip the API Gateway stuff from the template?  This is needed for HTTP to work.  You CANNOT remove the binary config from API Gateway.  API Gateway has to be explicitly told what the binary media types are and will ignore the isBase64Encoded flag if set.

**Comment by patriot1burke** [comment]
And why do you want to build native via docker?

**Comment by patriot1burke** [comment]
I just don't see any of this as improvements.

**Comment by oztimpower** [comment]
This is for the Amazon-Lambda extension, not the Amazon-Lambda-HTTP.  The difference being the later has API Gateway.

I think you may be confusing the earlier work, this work is to bring the Amazon-Lambda extension in line with the HTTP extension and add the SAM CLI

**Comment by patriot1burke** [comment]
Why do we want to use SAM if we don't have to?

**Comment by oztimpower** [comment]
Ha?  SAM is for local testing.

**Comment by patriot1burke** [comment]
Ya, but it sucks

**Comment by oztimpower** [comment]
Hmm, at least provide the documentation for people to test.

It's super simple and you have it on the HTTP extension, it is strange to have inconsistent docs.  With respect this is a big improvement over what was there.  Indeed it help resolve another issue submitted wherein a person was testing on Windows using the Docker Toolkit, and yet did not have an AWS Account.  There were path formatting issues on Windows and a PR was submitted.

**Comment by oztimpower** [comment]
For event based testing, which is what this is, SAM CLI is fast and efficient.  There is little complexity when compared to the API Gateway integration.

The issue I referred to which was resolved via the help of SAM (which is just Docker) is #5360.

**Comment by patriot1burke** [comment]
Apologies for the freakout, I'm change resistant (aka old).

Beyond the ridiculous .yaml soup, another big reason I don't like sam is that it is yet another thing you have to install.  The docs need to make clear that SAM is optional and that you don't have to install it to create a "Hello World" lambda.  The time it takes from project generation to deployment is important.  

The same is true about docker builds.  We do not want to force the developer to install docker if they don't have to.  The event driven, scale to zero is not very interesting to me.  What IS interesting to me is the reduction in deployment steps.

Also, we don't know if developers are adapting an existing project or not.   I like manage.sh, but the docs must at least break down the manage.sh script and describe what is going on, pointing to the relevant AWS CLI commands similarly how the docs go through the pom.xml changes required.

**Comment by oztimpower** [comment]
My initial feedback was that I liked the AWS CLI breakdown in the documentation too, as I'm a Gradle user and it helped me test my Lambda code without having to use Maven (and I have the same feelings towards Maven that you describe about SAM).

There are many reasons to test locally, I agree not all users will have Docker locally installed especially in a corporate environment where Docker for Desktop can be banned, yet for me I had many issues getting up and running with the native GRAALVM executable which I was able to solve using SAM, such as when vert.x was introduced and it required a /tmp folder, which would not work on AWS as it is /var/tmp, and I submitted a PR back to add -Djava.io.tmpfile=/tmp on the back of my SAM (Docker) tests.  

I did also for own purposes extract out the integration tests being used for the Amazon Lambda, that similarly may be an option for a vanilla (no side-effect) test without Docker for someone test locally.
